crates/noirc_frontend/src/hir/resolution/resolver.rs

```diff
@@ -835,6 +851,108 @@ impl<'a> Resolver<'a> {
+                self.try_eval_array_length_ident(ident.id, span)
+            }
+            ExpressionKind::Path(path) => {
+                let ident = self.get_ident_from_path(path.clone());
+                self.try_eval_array_length_ident(ident.id, span)
+            }
+            ExpressionKind::Prefix(operator) => {
+                let value = self.try_eval_array_length(&operator.rhs)?;
+                match operator.operator {
+                    crate::UnaryOp::Minus => Ok(0 - value),
+                    crate::UnaryOp::Not => Ok(!value),
+                }
+            }
+            ExpressionKind::Infix(operator) => {
+                let lhs = self.try_eval_array_length(&operator.lhs)?;
+                let rhs = self.try_eval_array_length(&operator.rhs)?;
+                match operator.operator.contents {
+                    crate::BinaryOpKind::Add => Ok(lhs + rhs),
+                    crate::BinaryOpKind::Subtract => Ok(lhs - rhs),
+                    crate::BinaryOpKind::Multiply => Ok(lhs * rhs),
+                    crate::BinaryOpKind::Divide => Ok(lhs / rhs),
+                    crate::BinaryOpKind::ShiftRight => Ok(lhs >> rhs),
+                    crate::BinaryOpKind::ShiftLeft => Ok(lhs << rhs),
+                    crate::BinaryOpKind::Modulo => Ok(lhs % rhs),
+                    crate::BinaryOpKind::And => Ok(lhs & rhs),
+                    crate::BinaryOpKind::Or => Ok(lhs | rhs),
+                    crate::BinaryOpKind::Xor => Ok(lhs ^ rhs),
+
+                    crate::BinaryOpKind::Equal
+                    | crate::BinaryOpKind::NotEqual
+                    | crate::BinaryOpKind::Less
+                    | crate::BinaryOpKind::LessEqual
+                    | crate::BinaryOpKind::Greater
+                    | crate::BinaryOpKind::GreaterEqual => {
+                        Err(Some(ResolverError::InvalidArrayLengthExpr { span }))
+                    }
+                }
+            }
+
+            ExpressionKind::Literal(_)
+            | ExpressionKind::Block(_)
+            | ExpressionKind::Index(_)
+            | ExpressionKind::Call(_)
+            | ExpressionKind::MethodCall(_)
+            | ExpressionKind::Constructor(_)
+            | ExpressionKind::MemberAccess(_)
+            | ExpressionKind::Cast(_)
+            | ExpressionKind::For(_)
+            | ExpressionKind::If(_)
+            | ExpressionKind::Tuple(_) => Err(Some(ResolverError::InvalidArrayLengthExpr { span })),
+
+            ExpressionKind::Error => Err(None),
+        }
+    }
+
+    fn try_eval_array_length_ident(
+        &mut self,
+        id: DefinitionId,
+        span: Span,
+    ) -> Result<u128, Option<ResolverError>> {
+        if id == DefinitionId::dummy_id() {
+            return Err(None); // error already reported
+        }
+
+        let definition = self.interner.definition(id);
+        match definition.rhs {
+            Some(rhs) if definition.is_global => self.try_eval_array_length_id(rhs),
```

Comment:
We don't have to restrict this to only globals right?

Reply:
We don't, but analysis for non-globals would need to be more thorough. E.g. even restricting to immutable variables we'd have to recursively check the definitions of these variables to make sure they are not parameters, or don't refer to parameters, or don't use variables whose definitions refer to parameters, etc.

Reply:
I am going to approve as this PR fulfills its scope. I am curious now though, should we be restricting to use only globals in `resolve_fixed_variable_array_length` as well then? Could you elaborate a bit more on this point?

Reply:
`resolve_fixed_variable_array_length` doesn't evaluate at all, it only checks if the rhs of an immutable variable is an integer literal with no recursion possible. It is when evaluating identifiers within this rhs we have to be careful of recursion and fail when the rhs is unknown or is a parameter.
This can be changed to follow identifiers and ensure they have an rhs (failing otherwise) fairly easily, it is just reporting this as an error to users may be confusing. If you try to use a variable which, e.g., depends on a variable which depends on a value computed in an if expression. It is the if expression that causes the error but this may be far away from the variable being used in the array expression.

Reply:
Ah ok this makes sense. I blanked that this lives within a mini-evaluator. Thank you for the clarification. The PR LGTM then, as adding this seems like a separate scope PR or for when we delay evaluation until SSA.